### ISSUE TITLE
Windows: conv_grad_ops_3d.cc and conv_ops_3d.cc take a long time to compile

### ISSUE THREAD

**Issue by meteorcloudy** [tensorflow/tensorflow#10521]
From http://ci.tensorflow.org/job/tf-master-win-bzl/1073/console, I see:
```
00:54:33 [3,344 / 3,367] Compiling tensorflow/core/kernels/conv_grad_ops_3d.cc; 906s standalone ... (2 actions running)
00:58:07 [3,344 / 3,367] Compiling tensorflow/core/kernels/conv_grad_ops_3d.cc; 1121s standalone ... (2 actions running)
01:01:55 [3,344 / 3,367] Compiling tensorflow/core/kernels/conv_grad_ops_3d.cc; 1348s standalone ... (2 actions running)
01:06:33 [3,344 / 3,367] Compiling tensorflow/core/kernels/conv_grad_ops_3d.cc; 1626s standalone ... (2 actions running)
01:12:26 [3,345 / 3,367] Compiling tensorflow/core/kernels/conv_ops_3d.cc; 1978s standalone
```
As you can see, it takes about half an hour to compile these two files.
This has been a bottleneck of the TF Windows Bazel build time for a while.
I guess it's because we are compiling with `/O2` option, so compiler was spending too much time on optimizing. But is there anything we can do to optimize the build time?

**Comment by gunan** [comment]
@mrry @guschmue @vit-stepanovs 
Any ideas?

**Comment by mrry** [comment]
No idea... perhaps @mjanusz (the original author) has some sense of why this code might take a long time to compile, but I don't see any extreme number of template instantiations that might contribute....

**Comment by guschmue** [comment]
just tried compiling for cpu - true, it takes long.
Trying to isolate this a little: if I take out the call to functor::CuboidConvolution<CPUDevice, T>()(...) its fine. Need to look a little what is behind that.

**Comment by guschmue** [comment]
I tried vs2017 - same thing. There are a bunch of references that excessive use of templates result in slowness and the code behind functor::CuboidConvolution uses a good amount of templates. Let me see if I can find help.

**Comment by guschmue** [comment]
found somebody to look at it; we update the issue once we know more.

**Comment by gunan** [comment]
@guschmue Any updates?

**Comment by guschmue** [comment]
I know the compiler team looked at it and there was a bug filed but it has been quite since. Let me ping them.

**Comment by rongjiecomputer** [comment]
Any improvement on this is welcomed. Currently `conv_grad_ops_3d.cc` and `conv_ops_3d.cc` took 2 hours to compile each on my machine (2 Core + 4GB RAM + HDD). I have to try to avoid making changes to common header files to avoid recompiling these two files...

**Comment by meteorcloudy** [comment]
@rongjiecomputer The slow compiling is caused by `__force_inline` used in eigen, we have a change to remove `__force_inline`, but it's not merged yet.

If you don't mind you can use @gunan's temporary repository. Follow the change in https://github.com/tensorflow/tensorflow/pull/15265.

**Comment by rongjiecomputer** [comment]
The build time improvement is huge! Less than 1 hour now.

I am a bit worried about the impact of runtime performance after the removal of `__force_inline` though.

**Comment by meteorcloudy** [comment]
Yes, that's why we haven't merged this, we want to make sure it doesn't hurt the performance too much by running some benchmark first.

**Comment by rongjiecomputer** [comment]
In V8 (Chrome), there are 3 build configurations: Debug, Release and Official build mode. Release mode turns on most of the optimizations so that final binary runs fast enough for most testing/benchmarking purpose, but the time-consuming build optimizations like LTO are only turned on in Official build.

Maybe we can have three modes as well, enable __force_inline in Official Build mode and not in Release mode. This will be very useful for external contributors.

**Comment by guschmue** [comment]
Some perf numbers for stock builds and and builds without __force_inline. 
There is some impact, maybe 10% on average - details below.
This is from a little benchmark tool that runs some simple fc, lstm, conv nets and takes numbers for training and eval. The eval numbers (ms units, the tool runs 1000 and takes the average) are worse than training.
The diff column shows the change compared to a stock build.

net | name | env | time(sec) | diff% | evaltime(ms) | diff%
-- | -- | -- | -- | -- | -- | --
conv | stock | vs2015.3 | 809.27 |   | 1.4375 |  
conv | nofi | vs2015.3 | 844.64 | 4.37060561 | 1.5937 | 10.86609
conv | stock | vs2017.15.4 | 783.12 |   | 1.4063 |  
conv | nofi | vs2017.15.4 | 824.65 | 5.30314639 | 1.5469 | 9.997867
fc | stock | vs2015.3 | 178.28 |   | 0.3281 |  
fc | nofi | vs2015.3 | 194.4 | 9.04195647 | 0.3906 | 19.04907
fc | stock | vs2017.15.4 | 180.66 |   | 0.3437 |  
fc | nofi | vs2017.15.4 | 185.48 | 2.66799513 | 0.3437 | 0
lstm | stock | vs2015.3 | 176.34 |   | 1.8906 |  
lstm | nofi | vs2015.3 | 202.7 | 14.9483951 | 2.2728 | 20.2158
lstm | stock | vs2017.15.4 | 182.29 |   | 1.9471 |  
lstm | nofi | vs2017.15.4 | 197.25 | 8.2067036 | 1.9063 | -2.09542

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by rongjiecomputer** [comment]
From https://blogs.msdn.microsoft.com/vcblog/2018/01/04/visual-studio-2017-throughput-improvements-and-advice/

> Beyond just turning off the optimizer for functions with pathological compile times, I also need to warn against too liberal a use of __forceinline when possible. ... [SNIP] ... This is the root cause of long compile times in places like Tensorflow/libsodium. This is something we are looking to address in a future release.

Good to know that at least Visual C++ team is aware of this issue.

**Comment by gunan** [comment]
Thanks to a lot of help from @guschmue we were able to debug this to use of `__forceinline` in eigen.
Unfortunately, removing `__forceinline` brings performance hits in the built binary, which is unacceptable for TF. Therefore, in our presubmits we remove forceinline, and our releases we do use forceinline.

Since this is a VS-eigen issue that unfortunately hits TF in a bad way, I will close this issue.
There is not much else we can do given all the information.
